// @wordpress/components/build/navigation/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NavigationMenu;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _constants = require("./constants");

var _context = require("./context");

var _navigationStyles = require("./styles/navigation-styles");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function NavigationMenu(_ref) {
  var backButtonLabel = _ref.backButtonLabel,
      children = _ref.children,
      className = _ref.className,
      _ref$menu = _ref.menu,
      menu = _ref$menu === void 0 ? _constants.ROOT_MENU : _ref$menu,
      parentMenu = _ref.parentMenu,
      title = _ref.title;

  var _useNavigationContext = (0, _context.useNavigationContext)(),
      activeMenu = _useNavigationContext.activeMenu,
      setActiveMenu = _useNavigationContext.setActiveMenu;

  if (activeMenu !== menu) {
    return null;
  }

  var classes = (0, _classnames.default)('components-navigation__menu', className);
  return (0, _element.createElement)("div", {
    className: classes
  }, parentMenu && (0, _element.createElement)(_navigationStyles.MenuBackButtonUI, {
    className: "components-navigation__back-button",
    isTertiary: true,
    onClick: function onClick() {
      return setActiveMenu(parentMenu, 'right');
    }
  }, (0, _element.createElement)(_icons.Icon, {
    icon: _icons.chevronLeft
  }), backButtonLabel || (0, _i18n.__)('Back')), (0, _element.createElement)(_navigationStyles.MenuUI, null, title && (0, _element.createElement)(_navigationStyles.MenuTitleUI, {
    as: "h2",
    className: "components-navigation__menu-title",
    variant: "subtitle"
  }, title), children));
}
