// @wordpress/components/build-module/navigation/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
import { noop } from 'lodash';
/**
 * WordPress dependencies
 */

import { useEffect, useRef, useState } from '@wordpress/element';
/**
 * Internal dependencies
 */

import Animate from '../animate';
import { ROOT_MENU } from './constants';
import { NavigationContext } from './context';
import { NavigationUI } from './styles/navigation-styles';
export default function Navigation(_ref) {
  var activeItem = _ref.activeItem,
      _ref$activeMenu = _ref.activeMenu,
      activeMenu = _ref$activeMenu === void 0 ? ROOT_MENU : _ref$activeMenu,
      children = _ref.children,
      className = _ref.className,
      _ref$onActivateItem = _ref.onActivateItem,
      onActivateItem = _ref$onActivateItem === void 0 ? noop : _ref$onActivateItem,
      _ref$onActivateMenu = _ref.onActivateMenu,
      onActivateMenu = _ref$onActivateMenu === void 0 ? noop : _ref$onActivateMenu;

  var _useState = useState(activeItem),
      _useState2 = _slicedToArray(_useState, 2),
      item = _useState2[0],
      setItem = _useState2[1];

  var _useState3 = useState(activeMenu),
      _useState4 = _slicedToArray(_useState3, 2),
      menu = _useState4[0],
      setMenu = _useState4[1];

  var _useState5 = useState(),
      _useState6 = _slicedToArray(_useState5, 2),
      slideOrigin = _useState6[0],
      setSlideOrigin = _useState6[1];

  var setActiveItem = function setActiveItem(itemId) {
    setItem(itemId);
    onActivateItem(itemId);
  };

  var setActiveMenu = function setActiveMenu(menuId) {
    var slideInOrigin = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'left';
    setSlideOrigin(slideInOrigin);
    setMenu(menuId);
    onActivateMenu(menuId);
  }; // Used to prevent the sliding animation on mount


  var isMounted = useRef(false);
  useEffect(function () {
    if (!isMounted.current) {
      isMounted.current = true;
    }
  }, []);
  useEffect(function () {
    if (activeItem !== item) {
      setActiveItem(activeItem);
    }

    if (activeMenu !== menu) {
      setActiveMenu(activeMenu);
    }
  }, [activeItem, activeMenu]);
  var context = {
    activeItem: item,
    activeMenu: menu,
    setActiveItem: setActiveItem,
    setActiveMenu: setActiveMenu
  };
  var classes = classnames('components-navigation', className);
  return createElement(NavigationUI, {
    className: classes
  }, createElement(Animate, {
    key: menu,
    type: "slide-in",
    options: {
      origin: slideOrigin
    }
  }, function (_ref2) {
    var animateClassName = _ref2.className;
    return createElement("div", {
      className: classnames(_defineProperty({}, animateClassName, isMounted.current && slideOrigin))
    }, createElement(NavigationContext.Provider, {
      value: context
    }, children));
  }));
}
