// @wordpress/block-library/build-module/legacy-widget/edit/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement, Fragment } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { get, omit } from 'lodash';
/**
 * WordPress dependencies
 */

import { Component } from '@wordpress/element';
import { Button, PanelBody, ToolbarGroup } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { withDispatch, withSelect } from '@wordpress/data';
import { BlockControls, InspectorControls } from '@wordpress/block-editor';
import ServerSideRender from '@wordpress/server-side-render';
import { update } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import LegacyWidgetEditHandler from './handler';
import LegacyWidgetPlaceholder from './placeholder';

var LegacyWidgetEdit = /*#__PURE__*/function (_Component) {
  _inherits(LegacyWidgetEdit, _Component);

  var _super = _createSuper(LegacyWidgetEdit);

  function LegacyWidgetEdit() {
    var _this;

    _classCallCheck(this, LegacyWidgetEdit);

    _this = _super.apply(this, arguments);
    _this.state = {
      hasEditForm: true,
      isPreview: false
    };
    _this.switchToEdit = _this.switchToEdit.bind(_assertThisInitialized(_this));
    _this.switchToPreview = _this.switchToPreview.bind(_assertThisInitialized(_this));
    _this.changeWidget = _this.changeWidget.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(LegacyWidgetEdit, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          attributes = _this$props.attributes,
          availableLegacyWidgets = _this$props.availableLegacyWidgets,
          hasPermissionsToManageWidgets = _this$props.hasPermissionsToManageWidgets,
          isSelected = _this$props.isSelected,
          prerenderedEditForm = _this$props.prerenderedEditForm,
          setAttributes = _this$props.setAttributes,
          setWidgetId = _this$props.setWidgetId,
          widgetId = _this$props.widgetId;
      var _this$state = this.state,
          isPreview = _this$state.isPreview,
          hasEditForm = _this$state.hasEditForm;
      var widgetClass = attributes.widgetClass;
      var widgetObject = widgetId && availableLegacyWidgets[widgetId] || widgetClass && availableLegacyWidgets[widgetClass];

      if (!widgetId && !widgetClass) {
        return createElement(LegacyWidgetPlaceholder, {
          availableLegacyWidgets: availableLegacyWidgets,
          hasPermissionsToManageWidgets: hasPermissionsToManageWidgets,
          onChangeWidget: function onChangeWidget(newWidget) {
            var _availableLegacyWidge = availableLegacyWidgets[newWidget],
                isReferenceWidget = _availableLegacyWidge.isReferenceWidget,
                idBase = _availableLegacyWidge.id_base;

            if (isReferenceWidget) {
              setWidgetId(newWidget);
            }

            setAttributes({
              instance: {},
              idBase: idBase,
              widgetClass: isReferenceWidget ? undefined : newWidget
            });
          }
        });
      }

      var inspectorControls = widgetObject ? createElement(InspectorControls, null, createElement(PanelBody, {
        title: widgetObject.name
      }, widgetObject.description)) : null;

      if (!hasPermissionsToManageWidgets) {
        return createElement(Fragment, null, inspectorControls, this.renderWidgetPreview());
      }

      return createElement(Fragment, null, createElement(BlockControls, null, createElement(ToolbarGroup, null, widgetObject && !widgetObject.isHidden && createElement(Button, {
        onClick: this.changeWidget,
        label: __('Change widget'),
        icon: update
      }), hasEditForm && createElement(Fragment, null, createElement(Button, {
        className: "components-tab-button",
        isPressed: !isPreview,
        onClick: this.switchToEdit
      }, createElement("span", null, __('Edit'))), createElement(Button, {
        className: "components-tab-button",
        isPressed: isPreview,
        onClick: this.switchToPreview
      }, createElement("span", null, __('Preview')))))), inspectorControls, hasEditForm && createElement(LegacyWidgetEditHandler, {
        isSelected: isSelected,
        isVisible: !isPreview,
        id: widgetId,
        idBase: attributes.idBase || widgetId,
        number: attributes.number,
        prerenderedEditForm: prerenderedEditForm,
        widgetName: get(widgetObject, ['name']),
        widgetClass: attributes.widgetClass,
        instance: attributes.instance,
        onFormMount: function onFormMount(formData) {
          // Function-based widgets don't come with an object of settings, only
          // with a pre-rendered HTML form. Extracting settings from that HTML
          // before this stage is not trivial (think embedded <script>). An alternative
          // proposed here serializes the form back into widget settings immediately after
          // it's mounted.
          if (!attributes.widgetClass) {
            _this2.props.setAttributes({
              instance: formData
            });
          }
        },
        onInstanceChange: function onInstanceChange(newInstance, newHasEditForm) {
          if (newInstance) {
            _this2.props.setAttributes({
              instance: newInstance
            });
          }

          if (newHasEditForm !== _this2.hasEditForm) {
            _this2.setState({
              hasEditForm: newHasEditForm
            });
          }
        }
      }), (isPreview || !hasEditForm) && this.renderWidgetPreview());
    }
  }, {
    key: "changeWidget",
    value: function changeWidget() {
      this.switchToEdit();
      this.props.setAttributes({
        instance: {},
        id: undefined,
        widgetClass: undefined
      });
      this.setState({
        hasEditForm: true
      });
    }
  }, {
    key: "switchToEdit",
    value: function switchToEdit() {
      this.setState({
        isPreview: false
      });
    }
  }, {
    key: "switchToPreview",
    value: function switchToPreview() {
      this.setState({
        isPreview: true
      });
    }
  }, {
    key: "renderWidgetPreview",
    value: function renderWidgetPreview() {
      var _this$props2 = this.props,
          widgetId = _this$props2.widgetId,
          attributes = _this$props2.attributes;
      return createElement(ServerSideRender, {
        className: "wp-block-legacy-widget__preview",
        block: "core/legacy-widget",
        attributes: _objectSpread({
          widgetId: widgetId
        }, omit(attributes, 'id'))
      });
    }
  }]);

  return LegacyWidgetEdit;
}(Component);

export default withSelect(function (select, _ref) {
  var clientId = _ref.clientId;
  var widgetId = select('core/edit-widgets').getWidgetIdForClientId(clientId);
  var widget = select('core/edit-widgets').getWidget(widgetId);
  var editorSettings = select('core/block-editor').getSettings();
  var availableLegacyWidgets = editorSettings.availableLegacyWidgets,
      hasPermissionsToManageWidgets = editorSettings.hasPermissionsToManageWidgets;
  return {
    hasPermissionsToManageWidgets: hasPermissionsToManageWidgets,
    availableLegacyWidgets: availableLegacyWidgets,
    widgetId: widgetId,
    prerenderedEditForm: widget ? widget.rendered_form : ''
  };
})(withDispatch(function (dispatch, _ref2) {
  var clientId = _ref2.clientId;
  return {
    setWidgetId: function setWidgetId(id) {
      dispatch('core/edit-widgets').setWidgetIdForClientId(clientId, id);
    }
  };
})(LegacyWidgetEdit));
