// @wordpress/editor/build/store/controls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.awaitNextStateChange = awaitNextStateChange;
exports.getRegistry = getRegistry;
exports.localAutosaveGet = localAutosaveGet;
exports.localAutosaveSet = localAutosaveSet;
exports.localAutosaveClear = localAutosaveClear;
exports.default = void 0;

var _data = require("@wordpress/data");

/**
 * WordPress dependencies
 */

/**
 * Returns a control descriptor signalling to subscribe to the registry and
 * resolve the control promise only when the next state change occurs.
 *
 * @return {Object} Control descriptor.
 */
function awaitNextStateChange() {
  return {
    type: 'AWAIT_NEXT_STATE_CHANGE'
  };
}
/**
 * Returns a control descriptor signalling to resolve with the current data
 * registry.
 *
 * @return {Object} Control descriptor.
 */


function getRegistry() {
  return {
    type: 'GET_REGISTRY'
  };
}
/**
 * Function returning a sessionStorage key to set or retrieve a given post's
 * automatic session backup.
 *
 * Keys are crucially prefixed with 'wp-autosave-' so that wp-login.php's
 * `loggedout` handler can clear sessionStorage of any user-private content.
 *
 * @see https://github.com/WordPress/wordpress-develop/blob/6dad32d2aed47e6c0cf2aee8410645f6d7aba6bd/src/wp-login.php#L103
 *
 * @param {string}  postId     Post ID.
 * @param {boolean} isPostNew  Whether post new.
 * @return {string}            sessionStorage key
 */


function postKey(postId, isPostNew) {
  return "wp-autosave-block-editor-post-".concat(isPostNew ? 'auto-draft' : postId);
}

function localAutosaveGet(postId, isPostNew) {
  return window.sessionStorage.getItem(postKey(postId, isPostNew));
}

function localAutosaveSet(postId, isPostNew, title, content, excerpt) {
  window.sessionStorage.setItem(postKey(postId, isPostNew), JSON.stringify({
    post_title: title,
    content: content,
    excerpt: excerpt
  }));
}

function localAutosaveClear(postId, isPostNew) {
  window.sessionStorage.removeItem(postKey(postId, isPostNew));
}

var controls = {
  AWAIT_NEXT_STATE_CHANGE: (0, _data.createRegistryControl)(function (registry) {
    return function () {
      return new Promise(function (resolve) {
        var unsubscribe = registry.subscribe(function () {
          unsubscribe();
          resolve();
        });
      });
    };
  }),
  GET_REGISTRY: (0, _data.createRegistryControl)(function (registry) {
    return function () {
      return registry;
    };
  }),
  LOCAL_AUTOSAVE_SET: function LOCAL_AUTOSAVE_SET(_ref) {
    var postId = _ref.postId,
        isPostNew = _ref.isPostNew,
        title = _ref.title,
        content = _ref.content,
        excerpt = _ref.excerpt;
    localAutosaveSet(postId, isPostNew, title, content, excerpt);
  }
};
var _default = controls;
exports.default = _default;
