// @wordpress/block-editor/build-module/components/inserter/block-patterns-tab.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { fromPairs } from 'lodash';
/**
 * WordPress dependencies
 */

import { useMemo, useCallback } from '@wordpress/element';
import { __, _x } from '@wordpress/i18n';
import { useAsyncList } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import InserterPanel from './panel';
import { searchItems } from './search-items';
import InserterNoResults from './no-results';
import usePatternsState from './hooks/use-patterns-state';
import BlockPatternList from '../block-patterns-list';

function BlockPatternsSearchResults(_ref) {
  var filterValue = _ref.filterValue,
      onInsert = _ref.onInsert;

  var _usePatternsState = usePatternsState(onInsert),
      _usePatternsState2 = _slicedToArray(_usePatternsState, 3),
      patterns = _usePatternsState2[0],
      onClick = _usePatternsState2[2];

  var currentShownPatterns = useAsyncList(patterns);
  var filteredPatterns = useMemo(function () {
    return searchItems(patterns, filterValue);
  }, [filterValue, patterns]);

  if (filterValue) {
    return !!filteredPatterns.length ? createElement(InserterPanel, {
      title: __('Search Results')
    }, createElement(BlockPatternList, {
      shownPatterns: currentShownPatterns,
      blockPatterns: filteredPatterns,
      onClickPattern: onClick
    })) : createElement(InserterNoResults, null);
  }
}

function BlockPatternsPerCategories(_ref2) {
  var onInsert = _ref2.onInsert;

  var _usePatternsState3 = usePatternsState(onInsert),
      _usePatternsState4 = _slicedToArray(_usePatternsState3, 3),
      patterns = _usePatternsState4[0],
      categories = _usePatternsState4[1],
      onClick = _usePatternsState4[2];

  var getPatternIndex = useCallback(function (pattern) {
    if (!pattern.categories || !pattern.categories.length) {
      return Infinity;
    }

    var indexedCategories = fromPairs(categories.map(function (_ref3, index) {
      var name = _ref3.name;
      return [name, index];
    }));
    return Math.min.apply(Math, _toConsumableArray(pattern.categories.map(function (category) {
      return indexedCategories[category] !== undefined ? indexedCategories[category] : Infinity;
    })));
  }, [categories]); // Ordering the patterns per category is important for the async rendering.

  var orderedPatterns = useMemo(function () {
    return patterns.sort(function (a, b) {
      return getPatternIndex(a) - getPatternIndex(b);
    });
  }, [patterns, getPatternIndex]);
  var currentShownPatterns = useAsyncList(orderedPatterns); // Uncategorized Patterns

  var uncategorizedPatterns = useMemo(function () {
    return patterns.filter(function (pattern) {
      return getPatternIndex(pattern) === Infinity;
    });
  }, [patterns, getPatternIndex]);
  return createElement(Fragment, null, categories.map(function (patternCategory) {
    var categoryPatterns = patterns.filter(function (pattern) {
      return pattern.categories && pattern.categories.includes(patternCategory.name);
    });
    return !!categoryPatterns.length && createElement(InserterPanel, {
      key: patternCategory.name,
      title: patternCategory.label
    }, createElement(BlockPatternList, {
      shownPatterns: currentShownPatterns,
      blockPatterns: categoryPatterns,
      onClickPattern: onClick
    }));
  }), !!uncategorizedPatterns.length && createElement(InserterPanel, {
    title: _x('Uncategorized')
  }, createElement(BlockPatternList, {
    shownPatterns: currentShownPatterns,
    blockPatterns: uncategorizedPatterns,
    onClickPattern: onClick
  })));
}

function BlockPatternsTabs(_ref4) {
  var onInsert = _ref4.onInsert,
      filterValue = _ref4.filterValue;
  return filterValue ? createElement(BlockPatternsSearchResults, {
    onInsert: onInsert,
    filterValue: filterValue
  }) : createElement(BlockPatternsPerCategories, {
    onInsert: onInsert
  });
}

export default BlockPatternsTabs;
