// @wordpress/editor/build-module/components/table-of-contents/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { Dropdown, Button } from '@wordpress/components';
import { useSelect } from '@wordpress/data';
import { info } from '@wordpress/icons';
import { forwardRef } from '@wordpress/element';
/**
 * Internal dependencies
 */

import TableOfContentsPanel from './panel';

function TableOfContents(_ref, ref) {
  var hasOutlineItemsDisabled = _ref.hasOutlineItemsDisabled,
      repositionDropdown = _ref.repositionDropdown,
      props = _objectWithoutProperties(_ref, ["hasOutlineItemsDisabled", "repositionDropdown"]);

  var hasBlocks = useSelect(function (select) {
    return !!select('core/block-editor').getBlockCount();
  }, []);
  return createElement(Dropdown, {
    position: repositionDropdown ? 'middle right right' : 'bottom',
    className: "table-of-contents",
    contentClassName: "table-of-contents__popover",
    renderToggle: function renderToggle(_ref2) {
      var isOpen = _ref2.isOpen,
          onToggle = _ref2.onToggle;
      return createElement(Button, _extends({}, props, {
        ref: ref,
        onClick: hasBlocks ? onToggle : undefined,
        icon: info,
        "aria-expanded": isOpen
        /* translators: button label text should, if possible, be under 16 characters. */
        ,
        label: __('Details'),
        tooltipPosition: "bottom",
        "aria-disabled": !hasBlocks
      }));
    },
    renderContent: function renderContent(_ref3) {
      var onClose = _ref3.onClose;
      return createElement(TableOfContentsPanel, {
        onRequestClose: onClose,
        hasOutlineItemsDisabled: hasOutlineItemsDisabled
      });
    }
  });
}

export default forwardRef(TableOfContents);
