// @wordpress/editor/build-module/components/global-keyboard-shortcuts/visual-editor-shortcuts.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useShortcut } from '@wordpress/keyboard-shortcuts';
import { useDispatch } from '@wordpress/data';
import deprecated from '@wordpress/deprecated';
import { BlockEditorKeyboardShortcuts } from '@wordpress/block-editor';
/**
 * Internal dependencies
 */

import SaveShortcut from './save-shortcut';

function VisualEditorGlobalKeyboardShortcuts() {
  var _useDispatch = useDispatch('core/editor'),
      redo = _useDispatch.redo,
      undo = _useDispatch.undo;

  useShortcut('core/editor/undo', function (event) {
    undo();
    event.preventDefault();
  }, {
    bindGlobal: true
  });
  useShortcut('core/editor/redo', function (event) {
    redo();
    event.preventDefault();
  }, {
    bindGlobal: true
  });
  return createElement(Fragment, null, createElement(BlockEditorKeyboardShortcuts, null), createElement(SaveShortcut, null));
}

export default VisualEditorGlobalKeyboardShortcuts;
export function EditorGlobalKeyboardShortcuts() {
  deprecated('EditorGlobalKeyboardShortcuts', {
    alternative: 'VisualEditorGlobalKeyboardShortcuts',
    plugin: 'Gutenberg'
  });
  return createElement(VisualEditorGlobalKeyboardShortcuts, null);
}
