// @wordpress/components/build-module/navigation/menu.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Icon, chevronLeft } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import { ROOT_MENU } from './constants';
import { useNavigationContext } from './context';
import { MenuBackButtonUI, MenuTitleUI, MenuUI } from './styles/navigation-styles';
export default function NavigationMenu(_ref) {
  var backButtonLabel = _ref.backButtonLabel,
      children = _ref.children,
      className = _ref.className,
      _ref$menu = _ref.menu,
      menu = _ref$menu === void 0 ? ROOT_MENU : _ref$menu,
      parentMenu = _ref.parentMenu,
      title = _ref.title;

  var _useNavigationContext = useNavigationContext(),
      activeMenu = _useNavigationContext.activeMenu,
      setActiveMenu = _useNavigationContext.setActiveMenu;

  if (activeMenu !== menu) {
    return null;
  }

  var classes = classnames('components-navigation__menu', className);
  return createElement("div", {
    className: classes
  }, parentMenu && createElement(MenuBackButtonUI, {
    className: "components-navigation__back-button",
    isTertiary: true,
    onClick: function onClick() {
      return setActiveMenu(parentMenu, 'right');
    }
  }, createElement(Icon, {
    icon: chevronLeft
  }), backButtonLabel || __('Back')), createElement(MenuUI, null, title && createElement(MenuTitleUI, {
    as: "h2",
    className: "components-navigation__menu-title",
    variant: "subtitle"
  }, title), children));
}
