// @wordpress/block-library/build-module/more/edit.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { PanelBody, ToggleControl } from '@wordpress/components';
import { useState } from '@wordpress/element';
import { InspectorControls } from '@wordpress/block-editor';
import { ENTER } from '@wordpress/keycodes';
import { getDefaultBlockName, createBlock } from '@wordpress/blocks';

var DEFAULT_TEXT = __('Read more');

export default function MoreEdit(_ref) {
  var _ref$attributes = _ref.attributes,
      customText = _ref$attributes.customText,
      noTeaser = _ref$attributes.noTeaser,
      insertBlocksAfter = _ref.insertBlocksAfter,
      setAttributes = _ref.setAttributes;

  var _useState = useState(DEFAULT_TEXT),
      _useState2 = _slicedToArray(_useState, 2),
      placeholder = _useState2[0],
      setPlaceholder = _useState2[1];

  var onChangeInput = function onChangeInput(event) {
    // Set defaultText to an empty string, allowing the user to clear/replace the input field's text
    setPlaceholder('');
    setAttributes({
      customText: event.target.value || undefined
    });
  };

  var onKeyDown = function onKeyDown(_ref2) {
    var keyCode = _ref2.keyCode;

    if (keyCode === ENTER) {
      insertBlocksAfter([createBlock(getDefaultBlockName())]);
    }
  };

  var getHideExcerptHelp = function getHideExcerptHelp(checked) {
    return checked ? __('The excerpt is hidden.') : __('The excerpt is visible.');
  };

  var toggleHideExcerpt = function toggleHideExcerpt() {
    return setAttributes({
      noTeaser: !noTeaser
    });
  };

  var value = customText !== null && customText !== void 0 ? customText : placeholder;
  var style = {
    width: "".concat(value.length + 1.2, "em")
  };
  return createElement(Fragment, null, createElement(InspectorControls, null, createElement(PanelBody, null, createElement(ToggleControl, {
    label: __('Hide the excerpt on the full content page'),
    checked: !!noTeaser,
    onChange: toggleHideExcerpt,
    help: getHideExcerptHelp
  }))), createElement("div", {
    className: "wp-block-more"
  }, createElement("input", {
    type: "text",
    value: value,
    onChange: onChangeInput,
    onKeyDown: onKeyDown,
    style: style
  })));
}
